sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java

```diff
@@ -0,0 +1,250 @@
+
+        // We generally put credential policy between BeforeRetry and AfterRetry policies and put Logging policy in the end.
+        // However since ACR uses the rest endpoints of the service in the credential policy,
+        // we want to be able to use the same pipeline (minus the credential policy) to have uniformity in the policy
+        // pipelines across all ACR endpoints.
+        if (credential == null) {
+            logger.verbose("Credentials are null, enabling anonymous access");
+        }
+
+        ArrayList<HttpPipelinePolicy> credentialPolicies = clone(policies);
+        credentialPolicies.add(loggingPolicy);
+
+        ContainerRegistryTokenService tokenService = new ContainerRegistryTokenService(
+            credential,
+            audience,
+            endpoint,
+            serviceVersion,
+            new HttpPipelineBuilder()
+                .policies(credentialPolicies.toArray(new HttpPipelinePolicy[0]))
+                .httpClient(httpClient)
+                .build(),
+            JacksonAdapter.createDefaultSerializerAdapter());
+
+        ContainerRegistryCredentialsPolicy credentialsPolicy = new ContainerRegistryCredentialsPolicy(tokenService);
+        policies.add(credentialsPolicy);
+
+        policies.add(loggingPolicy);
+        HttpPipeline httpPipeline =
+            new HttpPipelineBuilder()
+                .policies(policies.toArray(new HttpPipelinePolicy[0]))
+                .httpClient(httpClient)
+                .build();
+        return httpPipeline;
+    }
+
+    private static ArrayList<HttpPipelinePolicy> clone(ArrayList<HttpPipelinePolicy> policies) {
+        ArrayList<HttpPipelinePolicy> clonedPolicy = new ArrayList<>();
+        for (HttpPipelinePolicy policy:policies) {
+            clonedPolicy.add(policy);
+        }
+
+        return clonedPolicy;
+    }
+
+    /**
+     * This method computes the digest for the buffer content.
+     * Docker digest is a SHA256 hash of the docker image content and is deterministic based on the image build.
+     * @param buffer The buffer containing the image bytes.
+     * @return SHA-256 digest for the given buffer.
+     */
+    public static String computeDigest(ByteBuffer buffer) {
+        int startPosition = buffer.position();
+        try {
+            MessageDigest md = MessageDigest.getInstance("SHA-256");
+            md.update(buffer);
+            byte[] digest = md.digest();
+            return "sha256:" + byteArrayToHex(digest);
+
+        } catch (NoSuchAlgorithmException e) {
+        }
+        finally {
+            buffer.position(startPosition);
+        }
+
+        return null;
+    }
+
+    // TODO: Make this performant. We do not need String.format here and can potentially do simple bit manipulation.
+    public static String byteArrayToHex(byte[] a) {
+        StringBuilder sb = new StringBuilder(a.length * 2);
+        for (byte b: a) {
+            sb.append(String.format("%02x", b));
```

Comment:
Doing `String.format` in a tight loop like this will be fairly inefficient compared to writing out the byte-to-hex converter. We have this in a few places now and we should just make this a utility method in azure-core

Reply:
Yes, I have a TODO to fix this to use byte formatting instead, I have also created a tracking issue for this.
[https://github.com/Azure/azure-sdk-for-java/issues/27136](https://github.com/Azure/azure-sdk-for-java/issues/27136)